File: app/Sources/BKTree.swift
import Foundation

/// BK-tree over Hamming distance. Replaces the N² comparison matrix: at 100k assets
/// that would be 5 billion pairs.
///
/// Multi-valued on purpose — many files legitimately share one dHash (a burst, a
/// wallpaper pack), and a duplicate key must not become a duplicate node.
final class BKTree {
    private var keys: [UInt64] = []
    private var values: [[Int]] = []
    private var children: [[Int: Int]] = []
    private var root: Int?

    var count: Int { values.reduce(0) { $0 + $1.count } }
    var nodes: Int { keys.count }

    @inline(__always) static func distance(_ a: UInt64, _ b: UInt64) -> Int {
        (a ^ b).nonzeroBitCount
    }

    func add(_ key: UInt64, _ value: Int) {
        guard var cur = root else {
            keys.append(key); values.append([value]); children.append([:]); root = 0
            return
        }
        while true {
            let d = BKTree.distance(key, keys[cur])
            if d == 0 { values[cur].append(value); return }
            if let next = children[cur][d] { cur = next; continue }
            keys.append(key); values.append([value]); children.append([:])
            children[cur][d] = keys.count - 1
            return
        }
    }

    /// Every stored value whose key is within `radius` of `key`.
    func query(_ key: UInt64, radius: Int) -> [Int] {
        guard let root else { return [] }
        var out: [Int] = []
        var stack = [root]
        while let node = stack.popLast() {
            let d = BKTree.distance(key, keys[node])
            if d <= radius { out.append(contentsOf: values[node]) }
            let lo = d - radius, hi = d + radius
            for (edge, child) in children[node] where edge >= lo && edge <= hi {
                stack.append(child)
            }
        }
        return out
    }
}

/// Union-find, so a cluster is a connected component rather than a chain of pairs.
struct DisjointSet {
    private var parent: [Int]
    init(_ n: Int) { parent = Array(0..<n) }
    mutating func find(_ x: Int) -> Int {
        var r = x
        while parent[r] != r { parent[r] = parent[parent[r]]; r = parent[r] }
        return r
    }
    mutating func union(_ a: Int, _ b: Int) {
        let ra = find(a), rb = find(b)
        if ra != rb { parent[rb] = ra }
    }
}

File: cli/photomerge/bktree.py
"""A BK-tree over fixed-width hashes, queried by Hamming distance.

This replaces the v0 script's `bits[:,None,:] != bits[None,:,:]` (D1), which
materialises an N×N×64 boolean array: about 160 GB at 50,000 files, so the
process dies before it compares anything. A BK-tree stores one node per
*distinct* hash and answers a radius query by walking a few branches.

The structure works because Hamming distance is a metric, so the triangle
inequality holds: if `d(query, node)` is known, any subtree at distance `k` from
that node can only contain matches when `|k - d| <= radius`. Everything else is
pruned without being looked at.

Hashes arrive as bytes and are held as ints — `int.bit_count()` is a single
instruction's worth of work and beats unpacking bits into arrays.
"""

from __future__ import annotations

from collections import deque
from typing import Iterable


def to_int(digest: bytes) -> int:
    return int.from_bytes(digest, "big")


def hamming(a: int, b: int) -> int:
    return (a ^ b).bit_count()


class BKTree:
    """Multi-valued: several files routinely share one hash, and a duplicate
    hash must not become a duplicate node."""

    __slots__ = ("_keys", "_values", "_children", "_root")

    def __init__(self) -> None:
        self._keys: list[int] = []
        self._values: list[list] = []
        self._children: list[dict[int, int]] = []
        self._root: int | None = None

    def __len__(self) -> int:
        return sum(len(v) for v in self._values)

    @property
    def nodes(self) -> int:
        return len(self._keys)

    def add(self, key: int, value) -> None:
        if self._root is None:
            self._root = self._new(key, value)
            return
        node = self._root
        while True:
            distance = hamming(key, self._keys[node])
            if distance == 0:
                self._values[node].append(value)
                return
            child = self._children[node].get(distance)
            if child is None:
                self._children[node][distance] = self._new(key, value)
                return
            node = child

    def extend(self, items: Iterable[tuple[int, object]]) -> None:
        for key, value in items:
            self.add(key, value)

    def query(self, key: int, radius: int) -> list[tuple[int, list]]:
        """Every stored hash within `radius` of `key`, as `(distance, values)`."""
        if self._root is None:
            return []
        out: list[tuple[int, list]] = []
        stack = [self._root]
        while stack:
            node = stack.pop()
            distance = hamming(key, self._keys[node])
            if distance <= radius:
                out.append((distance, self._values[node]))
            # The triangle inequality: only these branches can hold a match.
            low, high = distance - radius, distance + radius
            for edge, child in self._children[node].items():
                if low <= edge <= high:
                    stack.append(child)
        return out

    def _new(self, key: int, value) -> int:
        self._keys.append(key)
        self._values.append([value])
        self._children.append({})
        return len(self._keys) - 1


def candidate_pairs(items: list[tuple[int, int]], radius: int) -> set[tuple[int, int]]:
    """`(file_id_a, file_id_b)` pairs whose hashes are within `radius`.

    Candidates only — PLAN §3.1-C. Nothing here is a merge; every pair still has
    to survive the verification cascade.
    """
    tree = BKTree()
    tree.extend((key, fid) for key, fid in items)
    pairs: set[tuple[int, int]] = set()
    for key, fid in items:
        for _, ids in tree.query(key, radius):
            for other in ids:
                if other != fid:
                    pairs.add((fid, other) if fid < other else (other, fid))
    return pairs


def connected_components(pairs: Iterable[tuple[int, int]]) -> list[set[int]]:
    """Group confirmed pairs into clusters.

    Deliberately *not* used on raw candidates: transitive closure over unverified
    near-matches is how a whole afternoon of similar photos becomes one asset.
    """
    adjacency: dict[int, set[int]] = {}
    for a, b in pairs:
        adjacency.setdefault(a, set()).add(b)
        adjacency.setdefault(b, set()).add(a)
    seen: set[int] = set()
    groups: list[set[int]] = []
    for start in adjacency:
        if start in seen:
            continue
        group = {start}
        seen.add(start)
        queue = deque([start])
        while queue:
            node = queue.popleft()
            for neighbour in adjacency[node]:
                if neighbour not in seen:
                    seen.add(neighbour)
                    group.add(neighbour)
                    queue.append(neighbour)
        groups.append(group)
    return groups
